prep release: v1.61.5 (#7348)
apollographql/router: CHANGELOG.md

This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

# [1.61.5] - 2025-04-28

> ## 🚀 Features

Author: ```suggestion
## 🔍 Debuggability
```

> ### Compute job spans ([PR #7236](https://github.com/apollographql/router/pull/7236))

Reviewer: Standardise on "compute job pool"
```suggestion
### Add compute job pool spans ([PR #7236](https://github.com/apollographql/router/pull/7236))
```

> The router uses a separate thread pool called "compute jobs" to ensure that requests do not block tokio io worker threads.

Reviewer: Standard compute job pool introduction
```suggestion
The compute job pool in the router is used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.
```

This PR adds spans to jobs that are on this pool to allow users to see when latency is introduced due to 
resource contention within the compute job pool.

> * `compute_job`:
>   - `job.type`: (`QueryParsing`|`QueryParsing`|`Introspection`)
> * `compute_job.execution`
>   - `job.age`: `P1`-`P8`
>   - `job.type`: (`QueryParsing`|`QueryParsing`|`Introspection`)

Reviewer: ```suggestion
* `compute_job`:
  - `job.type`: (`query_parsing`|`query_planning`|`introspection`)
* `compute_job.execution`
  - `job.age`: `P1`-`P8`
  - `job.type`: (`query_parsing`|`query_planning`|`introspection`)
```

Reviewer: (...Or did we not change this?)

Jobs are executed highest priority (`P8`) first. Jobs that are low priority (`P1`) age over time, eventually executing 
at highest priority. The age of a job is can be used to diagnose if a job was waiting in the queue due to other higher 
priority jobs also in the queue.

By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/7236

> ### Add compute pool metrics ([PR #7184](https://github.com/apollographql/router/pull/7184))

Reviewer: Standardise on "compute job pool"
```suggestion
### Add compute job pool metrics ([PR #7184](https://github.com/apollographql/router/pull/7184))
```

> The compute job pool is used within the router for compute intensive jobs that should not block the Tokio worker threads.

Reviewer: Standard compute job pool introduction
```suggestion
The compute job pool in the router is used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.
```

- `apollo.router.compute_jobs.duration` - A histogram of time spent in the compute pipeline by the job, including the queue and query planning.
  - `job.type`: (`query_planning`, `query_parsing`, `introspection`)
  - `job.outcome`: (`executed_ok`, `executed_error`, `channel_error`, `rejected_queue_full`, `abandoned`)
- `apollo.router.compute_jobs.queue.wait.duration` - A histogram of time spent in the compute queue by the job.
  - `job.type`: (`query_planning`, `query_parsing`, `introspection`)
- `apollo.router.compute_jobs.execution.duration` - A histogram of time spent to execute job (excludes time spent in the queue).
  - `job.type`: (`query_planning`, `query_parsing`, `introspection`)
- `apollo.router.compute_jobs.active_jobs` - A gauge of the number of compute jobs being processed in parallel.
  - `job.type`: (`query_planning`, `query_parsing`, `introspection`)

By [@carodewig](https://github.com/carodewig) in https://github.com/apollographql/router/pull/7184

## 🐛 Fixes

### Poll pending compute jobs hang request ([PR #7273](https://github.com/apollographql/router/pull/7273))

> Compute jobs in the router are used to execute CPU intensive work outside of the main io worker threads, in particular for QueryParsing, QueryPlanning and Introspection.

Author: Compute jobs in the router are used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.

> Compute jobs in the router are used to execute CPU intensive work outside of the main io worker threads, in particular for QueryParsing, QueryPlanning and Introspection.

Author: ```suggestion
Compute jobs in the router are used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.
```

> ### Poll pending compute jobs hang request ([PR #7273](https://github.com/apollographql/router/pull/7273))
> 
> Compute jobs in the router are used to execute CPU intensive work outside of the main io worker threads, in particular for QueryParsing, QueryPlanning and Introspection.
> 
> We previously checked if the compute job queue was full and returned `Poll::Pending` in the tower services.
> However, this will cause requests to hang until timeout.
> 
> This PR shifts the logic into `call` and will immediately return a `SERVICE_UNAVAILABLE` response to the user.

Reviewer: Focusing on user impact
```suggestion
### Fix hanging requests when compute job queue is full ([PR #7273](https://github.com/apollographql/router/pull/7273))

Compute jobs in the router are used to execute CPU intensive work outside of the main io worker threads, in particular for QueryParsing, QueryPlanning and Introspection.

When the compute job queue was full, requests could hang until timeout. Now, the router immediately returns a `SERVICE_UNAVAILABLE` response to the user.
```

> ### Poll pending compute jobs hang request ([PR #7273](https://github.com/apollographql/router/pull/7273))
> 
> Compute jobs in the router are used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.
> 
> We previously checked if the compute job queue was full and returned `Poll::Pending` in the tower services.
> However, this will cause requests to hang until timeout.
> 
> This PR shifts the logic into `call` and will immediately return a `SERVICE_UNAVAILABLE` response to the user.

Reviewer: Focusing on user impact
```suggestion
### Fix hanging requests when compute job queue is full ([PR #7273](https://github.com/apollographql/router/pull/7273))

The compute job pool in the router is used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection. When the pool is busy, jobs enter a queue.

When the compute job queue was full, requests could hang until timeout. Now, the router immediately returns a `SERVICE_UNAVAILABLE` response to the user.
```

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/7273

> ### Increase compute job worker pool queue size ([PR #7205](https://github.com/apollographql/router/pull/7205))

Reviewer: Standardise on "compute job pool"
```suggestion
### Increase compute job pool queue size ([PR #7205](https://github.com/apollographql/router/pull/7205))
```

> The compute job worker pool is used for CPU-bound tasks, like GraphQL parsing, validation, and query planning. When there are too many jobs to handle in parallel, jobs enter a queue.

Author: ```suggestion
Compute jobs in the router are used to execute CPU intensive work outside of the main io worker threads, in particular for QueryParsing, QueryPlanning and Introspection. When there are too many jobs to handle in parallel, jobs enter a queue.
```

> The compute job worker pool is used for CPU-bound tasks, like GraphQL parsing, validation, and query planning. When there are too many jobs to handle in parallel, jobs enter a queue.

Author: ```suggestion
Compute jobs in the router are used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.
```

> Compute jobs in the router are used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection.

Reviewer: Standardise on "compute job pool" + introduce the queue that the next line refers to
```suggestion
The compute job pool in the router is used to execute CPU intensive work outside of the main I/O worker threads, including GraphQL parsing, query planning, and introspection. When the pool is busy, jobs enter a queue.
```

We previously set this queue size to 20 (per thread). However, this may be too small on resource constrained environments.

> This patch increases the queue size to 1 000 jobs per thread. For reference, in older router versions before the introduction of the compute job worker pool, the equivalent queue size was *1000*.

Reviewer: US separator & correct old number:
```suggestion
This patch increases the queue size to 1,000 jobs per thread. For reference, in older router versions before the introduction of the compute job worker pool, the equivalent queue size was *10,000*.
```
Maybe it's too much detail? I wanted to list the old higher number to ease potential concerns that 1 000 is a lot

> This patch increases the queue size to 1,000 jobs per thread. For reference, in older router versions before the introduction of the compute job worker pool, the equivalent queue size was *10,000*.

Author: ```suggestion
This patch increases the queue size to 1,000 jobs per thread. For reference, in older router versions before the introduction of the compute job worker pool, the equivalent queue size was *1,000*.
```
